File: backend/files.py
import json
import logging
import os
import subprocess

from flask import Blueprint, request
from flask_login import current_user

from .classrooms import CLASSROOMS_JSON_FILE
from .docker import CLASSROOMS_ROOT, CONTAINER_USER_GID, CONTAINER_USER_UID

files_bp = Blueprint("upload", __name__)


def _translate_container_path(path: str) -> str:
    """Convert absolute container classroom paths to host paths."""
    if not path:
        return path
    if path.startswith("/classrooms/"):
        tail = path[len("/classrooms/") :].lstrip("/")
        return os.path.join(CLASSROOMS_ROOT, tail)
    return path


def set_container_ownership(path):
    """Set ownership of a file/directory to match the container user"""
    try:
        os.chown(path, CONTAINER_USER_UID, CONTAINER_USER_GID)
        # Also ensure proper permissions
        if os.path.isdir(path):
            os.chmod(path, 0o755)  # drwxr-xr-x for directories
        else:
            os.chmod(path, 0o644)  # -rw-r--r-- for files
    except OSError as e:
        # Log the error but don't fail - this might happen in dev environments
        logger = logging.getLogger("files")
        logger.warning(f"Failed to set ownership for {path}: {e}")


def _resolve_classroom_path(upload_dir: str, rel_path: str) -> str | None:
    """If rel_path begins with a symlink that targets /classrooms/<...>,
    map it to the corresponding host path under CLASSROOMS_ROOT.

    Returns the mapped absolute host path, or None if not a classroom symlink path.
    """
    try:
        # Normalize relative path
        safe_rel = (rel_path or "").lstrip("/")
        if not safe_rel:
            return None
        parts = safe_rel.split("/", 1)
        top = parts[0]
        remainder = parts[1] if len(parts) > 1 else ""
        top_abs = os.path.join(upload_dir, top)
        if not os.path.islink(top_abs):
            return None
        target = os.readlink(top_abs)
        if target.startswith("/classrooms/"):
            tail = target[len("/classrooms/") :].lstrip("/")
        elif target.startswith(CLASSROOMS_ROOT):
            tail = target[len(CLASSROOMS_ROOT) :].lstrip("/")
        else:
            return None
        host_base = os.path.join(CLASSROOMS_ROOT, tail)
        candidate = os.path.join(host_base, remainder) if remainder else host_base

        try:
            resolved = os.path.realpath(candidate)
        except OSError:
            resolved = candidate

        translated = _translate_container_path(resolved)
        normalized = os.path.normpath(translated)

        # Ensure the final path stays within the classrooms root
        try:
            if os.path.commonpath([CLASSROOMS_ROOT, normalized]) != CLASSROOMS_ROOT:
                return None
        except ValueError:
            return None

        return normalized
    except OSError:
        return None


def _append_classroom_tree_entries(
    file_list: list[str], slug_name: str, host_base: str, _visited_paths=None
):
    """Append directory and file entries for a classroom mount under the given slug name."""
    if not os.path.isdir(host_base):
        return

    # Track visited paths to avoid infinite symlink loops
    if _visited_paths is None:
        _visited_paths = set()

    # Resolve to canonical path to detect loops
    try:
        canonical = os.path.realpath(host_base)
        if canonical in _visited_paths:
            return
        _visited_paths.add(canonical)
    except OSError:
        return

    # Root of the classroom under this slug
    file_list.append(f"{slug_name}/")

    # First pass: handle symlinks specially at this level
    try:
        for entry in os.listdir(host_base):
            full_path = os.path.join(host_base, entry)
            if os.path.islink(full_path):
                try:
                    # Check if symlink target exists and is a directory
                    target = os.readlink(full_path)
                    # Try to resolve container paths to host paths
                    if target.startswith("/classrooms/"):
                        target_host = target.replace(
                            "/classrooms/", f"{CLASSROOMS_ROOT}/"
                        )
                    else:
                        # Relative or other absolute path
                        if not os.path.isabs(target):
                            target_host = os.path.join(host_base, target)
                        else:
                            target_host = target

                    # Check if it points to a directory
                    if os.path.isdir(target_host):
                        # Add as directory
                        dir_entry = f"{slug_name}/{entry}/"
                        if dir_entry not in file_list:
                            file_list.append(dir_entry)
                            # Recursively add contents
                            _append_classroom_tree_entries(
                                file_list,
                                f"{slug_name}/{entry}",
                                target_host,
                                _visited_paths,
                            )
                    else:
                        # Symlink to file
                        file_entry = f"{slug_name}/{entry}"
                        if file_entry not in file_list:
                            file_list.append(file_entry)
                except (OSError, IOError) as e:
                    logger = logging.getLogger("files")
                    logger.debug(f"Failed to process symlink {full_path}: {e}")
    except Exception as e:
        logger = logging.getLogger("files")
        logger.warning(f"Failed to process top-level symlinks in {host_base}: {e}")

    # Second pass: use os.walk for regular files and directories (skip symlinks)
    for root, dirs, files in os.walk(host_base, followlinks=False):
        rel = os.path.relpath(root, host_base)
        prefix = slug_name if rel == "." else f"{slug_name}/{rel}"

        # Filter out symlinks from dirs since we already handled them above
        for d in list(dirs):
            full_path = os.path.join(root, d)
            if not os.path.islink(full_path):
                entry = f"{prefix}/{d}/"
                if entry not in file_list:
                    file_list.append(entry)
            else:
                # Remove from dirs to prevent os.walk from descending
                dirs.remove(d)

        # Add regular files (not symlinks)
        for name in files:
            full_path = os.path.join(root, name)
            if not os.path.islink(full_path):
                entry = f"{prefix}/{name}"
                if entry not in file_list:
                    file_list.append(entry)


@files_bp.post("/upload")
def upload():
    if not current_user.is_authenticated:
        return "Unauthorized", 401

    user_id = current_user.id
    upload_dir = f"/tmp/uploads/{user_id}"
    os.makedirs(upload_dir, exist_ok=True)
    set_container_ownership(upload_dir)

    files = request.files.getlist("files")
    if not files:
        return "No files provided", 400
    for f in files:
        file_path = os.path.join(upload_dir, f.filename)
        f.save(file_path)
        set_container_ownership(file_path)

    return "File uploaded successfully", 200


@files_bp.post("/upload-folder")
def upload_folder():
    if not current_user.is_authenticated:
        return "Unauthorized", 401

    user_id = current_user.id
    upload_dir = f"/tmp/uploads/{user_id}"
    os.makedirs(upload_dir, exist_ok=True)
    set_container_ownership(upload_dir)

    files = request.files.getlist("files")
    if not files:
        return "No files provided", 400

    for f in files:
        safe_path = os.path.normpath(os.path.join(upload_dir, f.filename))
        if not safe_path.startswith(upload_dir):
            continue  # prevent directory traversal
        # Resolve classroom symlink destinations to host paths if needed
        classroom_target = _resolve_classroom_path(upload_dir, f.filename)
        dest_path = classroom_target if classroom_target else safe_path
        dir_path = os.path.dirname(dest_path)
        os.makedirs(dir_path, exist_ok=True)
        set_container_ownership(dir_path)
        f.save(dest_path)
        set_container_ownership(dest_path)

    # Check if move-into parameter is provided
    move_into = request.form.get("move-into")
    if move_into:
        # Execute an Enter before the cd command in the user's container using tmux send-keys
        container_name = f"user-container-{user_id}"
        try:
            # First, send an Enter key
            subprocess.run(
                [
                    "docker",
                    "exec",
                    container_name,
                    "tmux",
                    "send-keys",
                    "-t",
                    "3compute",
                    "Enter",
                ],
                check=True,
            )
            # Then, send the cd command
            subprocess.run(
                [
                    "docker",
                    "exec",
                    container_name,
                    "tmux",
                    "send-keys",
                    "-t",
                    "3compute",
                    f"cd /app/{move_into}",
                    "Enter",
                ],
                check=True,
            )
        except subprocess.CalledProcessError:
            # If the directory doesn't exist or cd fails, continue without error
            pass

    return "Folder uploaded successfully", 200


@files_bp.get("/list-files")
def list_files():
    if not current_user.is_authenticated:
        return "Unauthorized", 401

    user_id = current_user.id
    upload_dir = f"/tmp/uploads/{user_id}"

    if not os.path.exists(upload_dir):
        return {"files": []}, 200

    file_tree = []
    expanded_symlinks: set[str] = set()
    classroom_symlinks: dict[str, str] = {}

    logger = logging.getLogger("files")
    logger.debug(f"[{user_id}] Listing files in {upload_dir}")

    top_level_symlinks: set[str] = set()

    # Detect top-level classroom symlinks first so the backend knows about them
    try:
        for entry in os.listdir(upload_dir):
            full_path = os.path.join(upload_dir, entry)
            if not os.path.islink(full_path):
                continue

            target = os.readlink(full_path)
            logger.debug(f"[{user_id}] Found top-level symlink: {entry} -> {target}")

            tail = None
            if target.startswith(CLASSROOMS_ROOT):
                tail = target[len(CLASSROOMS_ROOT) :].lstrip("/")
            elif target.startswith("/classrooms/"):
                tail = target[len("/classrooms/") :].lstrip("/")
            else:
                logger.debug(
                    f"[{user_id}] Symlink {entry} does not target classrooms directory, skipping"
                )
                continue

            if not tail:
                continue

            host_base = os.path.join(CLASSROOMS_ROOT, tail)
            if not os.path.exists(host_base):
                logger.warning(
                    f"[{user_id}] Host classroom path missing for symlink {entry}: {host_base}"
                )
                continue

            _append_classroom_tree_entries(file_tree, entry, host_base)
            expanded_symlinks.add(f"{entry}/")
            top_level_symlinks.add(entry)

            classroom_id = tail.split("/", 1)[0]
            logger.debug(
                f"[{user_id}] Recording classroom symlink: {entry} -> {classroom_id}"
            )
            classroom_symlinks[entry] = classroom_id
    except FileNotFoundError:
        pass

    # Include user files under /tmp/uploads/<id>
    for root, dirs, files in os.walk(upload_dir):
        # Include directories (with trailing slash) so empty folders are discoverable
        for d in list(dirs):
            full_path = os.path.join(root, d)
            relative_path = os.path.relpath(full_path, upload_dir)

            # Skip top-level classroom symlinks already processed
            if relative_path.split(os.sep)[0] in top_level_symlinks:
                dirs.remove(d)
                continue

            # If this directory is a symlink to a classroom, expand its tree from the host path
            if os.path.islink(full_path):
                target = os.readlink(full_path)
                logger.debug(f"[{user_id}] Found symlink: {relative_path} -> {target}")
                if target.startswith("/classrooms/"):
                    tail = target[len("/classrooms/") :].lstrip("/")
                    host_base = os.path.join(CLASSROOMS_ROOT, tail)
                    _append_classroom_tree_entries(file_tree, relative_path, host_base)
                    expanded_symlinks.add(f"{relative_path}/")
                    if "/" not in relative_path:
                        classroom_id = tail.split("/", 1)[0]
                        logger.debug(
                            f"[{user_id}] Recording classroom symlink: {relative_path} -> {classroom_id}"
                        )
                        classroom_symlinks[relative_path] = classroom_id
                    # Prevent os.walk from descending into this symlink path
                    dirs.remove(d)
                    continue
            file_tree.append(f"{relative_path}/")

        # Include files
        for name in files:
            full_path = os.path.join(root, name)
            relative_path = os.path.relpath(full_path, upload_dir)
            if os.path.islink(full_path):
                target = os.readlink(full_path)
                if target.startswith("/classrooms/"):
                    tail = target[len("/classrooms/") :].lstrip("/")
                    host_base = os.path.join(CLASSROOMS_ROOT, tail)
                    _append_classroom_tree_entries(file_tree, relative_path, host_base)
                    expanded_symlinks.add(relative_path)
                    continue
            file_tree.append(relative_path)

    if expanded_symlinks:
        file_tree = [entry for entry in file_tree if entry not in expanded_symlinks]

    classroom_meta: dict[str, dict] = {}
    logger.debug(f"[{user_id}] Found classroom_symlinks: {classroom_symlinks}")

    if classroom_symlinks:
        all_classrooms = {}
        try:
            if os.path.exists(CLASSROOMS_JSON_FILE):
                with open(CLASSROOMS_JSON_FILE, "r") as f:
                    all_classrooms = json.load(f)
                logger.debug(
                    f"[{user_id}] Loaded classrooms.json with {len(all_classrooms)} classrooms"
                )
        except Exception as e:
            logger.warning(f"[{user_id}] Failed loading classrooms metadata: {e}")

        for slug, cid in classroom_symlinks.items():
            info = {}
            if isinstance(all_classrooms, dict):
                info = all_classrooms.get(cid, {}) or {}
            classroom_meta[slug] = {
                "id": cid,
                "name": info.get("name"),
                "archived": info.get("archived", False),
            }
            logger.debug(
                f"[{user_id}] Added classroom_meta[{slug}] = {classroom_meta[slug]}"
            )

    logger.debug(f"[{user_id}] Returning classroom_meta: {classroom_meta}")
    return {"files": file_tree, "classroomMeta": classroom_meta}, 200


@files_bp.post("/move")
def move_file_or_folder():
    """Move/rename a file or directory within the user's workspace.

    Expects JSON body: { "source": "/path/from", "destination": "/path/to" }
    Both paths are relative to the user's upload directory. Leading slashes are allowed.
    """
    if not current_user.is_authenticated:
        return "Unauthorized", 401

    data = request.get_json(silent=True) or {}
    source_param = (data.get("source") or "").lstrip("/")
    destination_param = (data.get("destination") or "").lstrip("/")
    overwrite = bool(data.get("overwrite"))

    if not source_param or not destination_param:
        return "Invalid path", 400

    user_id = current_user.id
    upload_dir = f"/tmp/uploads/{user_id}"

    src_path = os.path.normpath(os.path.join(upload_dir, source_param))
    dst_path = os.path.normpath(os.path.join(upload_dir, destination_param))

    # Security: ensure resulting paths stay within the user's directory
    if not src_path.startswith(upload_dir) or not dst_path.startswith(upload_dir):
        return "Invalid path", 400

    if not os.path.exists(src_path):
        return "Source not found", 404

    # Disallow moving a folder into itself or its descendants
    try:
        src_rel = os.path.relpath(src_path, upload_dir)
        dst_rel = os.path.relpath(dst_path, upload_dir)
        if dst_rel == src_rel or dst_rel.startswith(src_rel + os.sep):
            return {
                "error": "Cannot move a folder into itself or its subdirectory"
            }, 400
    except ValueError:
        # relpath can throw on different drives; unlikely in Linux containers, but guard anyway
        pass

    # Ensure destination parent exists
    dst_parent = os.path.dirname(dst_path)
    if dst_parent and not os.path.exists(dst_parent):
        try:
            os.makedirs(dst_parent, exist_ok=True)
            set_container_ownership(dst_parent)
        except NotADirectoryError:
            return {"error": "A file exists in the destination path"}, 409

    # Prevent overwriting existing files/folders unless overwrite flag is set
    if os.path.exists(dst_path):
        if not overwrite:
            return {"error": "Destination already exists"}, 409
        # If overwriting, remove existing destination first
        try:
            if os.path.isdir(dst_path):
                import shutil

                shutil.rmtree(dst_path)
            else:
                os.remove(dst_path)
        except OSError as e:
            return {"error": f"Failed to replace destination: {e}"}, 500

    try:
        os.rename(src_path, dst_path)
        # Set ownership on the moved item (best-effort)
        set_container_ownership(dst_path)
    except OSError as e:
        return {"error": f"Failed to move: {e}"}, 500

    return "Moved successfully", 200


@files_bp.route("/file/<path:filename>", methods=["GET", "PUT", "DELETE", "POST"])
def handle_file(filename):
    if not current_user.is_authenticated:
        return "Unauthorized", 401

    user_id = current_user.id
    upload_dir = f"/tmp/uploads/{user_id}"
    # Map classroom symlink paths to host filesystem if applicable
    mapped = _resolve_classroom_path(upload_dir, filename)
    file_path = mapped if mapped else os.path.join(upload_dir, filename)
    file_path = os.path.normpath(file_path)

    # Guard against breaking out of upload_dir / classrooms root
    try:
        allowed_roots = [upload_dir, CLASSROOMS_ROOT]
        if not any(
            os.path.commonpath([root, file_path]) == root for root in allowed_roots
        ):
            return "Invalid path", 400
    except ValueError:
        return "Invalid path", 400

    # Prevent writes to templates folders (read-only for participants)
    # Templates can only be modified by instructors via the dedicated upload endpoint
    if request.method in ("POST", "PUT", "DELETE"):
        # Check if path is in a templates folder (either directly or via classroom-templates symlink)
        rel_to_classrooms = ""
        if file_path.startswith(CLASSROOMS_ROOT):
            rel_to_classrooms = file_path[len(CLASSROOMS_ROOT):].lstrip("/")
        is_templates_path = (
            "/templates/" in f"/{rel_to_classrooms}" or
            rel_to_classrooms.endswith("/templates") or
            "/classroom-templates/" in filename or
            filename.startswith("classroom-templates/")
        )
        if is_templates_path:
            return {"error": "Templates folder is read-only. Use 'Copy to Workspace' instead."}, 403

    if request.method == "POST":
        # Create a new directory or file
        if filename.endswith("/"):
            # It's a directory
            # Normalize: avoid trailing slash for existence checks
            normalized_path = file_path[:-1]
            # If a file exists with the same name, return conflict
            if os.path.isfile(normalized_path):
                return {"error": "A file with the same name already exists"}, 409
            try:
                os.makedirs(file_path, exist_ok=True)
            except NotADirectoryError:
                # One of the parents is a file; cannot create directory path
                return {
                    "error": "A file exists in the path; cannot create directory"
                }, 409
            set_container_ownership(file_path)
            return "Directory created successfully", 200
        else:
            # It's a file - check if the directory exists, if not, create it
            dir_path = os.path.dirname(file_path)
            if dir_path and dir_path != upload_dir:
                os.makedirs(dir_path, exist_ok=True)
                set_container_ownership(dir_path)
        # create an empty file if it doesn't exist
        # If a directory with the same name exists, return conflict
        if os.path.isdir(file_path):
            return {"error": "A folder with the same name already exists"}, 409
        if os.path.exists(file_path):
            return "File already exists", 400
        with open(file_path, "w") as f:
            f.write("")
            set_container_ownership(file_path)
        return "File created successfully", 200

    elif not os.path.exists(file_path):
        return "File not found", 404

    elif request.method == "GET":
        image_extensions = ["jpg", "jpeg", "png", "gif", "bmp", "webp", "svg", "ico"]
        file_ext = filename.split(".")[-1].lower() if "." in filename else ""

        if file_ext in image_extensions:
            from flask import send_file

            return send_file(file_path)
        else:
            try:
                with open(file_path, "r", encoding="utf-8") as f:
                    content = f.read()
                return content, 200
            except UnicodeDecodeError:
                from flask import send_file

                return send_file(file_path)

    elif request.method == "PUT":
        content = request.data.decode("utf-8")
        with open(file_path, "w") as f:
            f.write(content)
        set_container_ownership(file_path)
        return "File updated successfully", 200

    elif request.method == "DELETE":
        if os.path.isdir(file_path):
            import shutil

            shutil.rmtree(file_path)
        else:
            os.remove(file_path)
        return "File deleted successfully", 200

File: go-backend/internal/files/files.go
package files

import (
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"3compute-backend/internal/auth"
	"3compute-backend/internal/docker"

	"github.com/gin-gonic/gin"
	"github.com/sirupsen/logrus"
)

// RegisterRoutes registers file-related routes
func RegisterRoutes(r *gin.Engine) {
	r.POST("/upload", auth.RequireAuth(), handleUpload)
	r.POST("/upload-folder", auth.RequireAuth(), handleUploadFolder)
	r.GET("/list-files", auth.RequireAuth(), handleListFiles)
	r.POST("/move", auth.RequireAuth(), handleMoveFileOrFolder)
	r.Match([]string{"GET", "PUT", "DELETE", "POST"}, "/file/*filepath", auth.RequireAuth(), handleFile)
}

// setContainerOwnership sets ownership of a file/directory to match the container user
func setContainerOwnership(path string) {
	if err := os.Chown(path, docker.ContainerUserUID, docker.ContainerUserGID); err != nil {
		logrus.WithError(err).Warnf("Failed to set ownership for %s", path)
		return
	}

	// Also ensure proper permissions
	info, err := os.Stat(path)
	if err != nil {
		logrus.WithError(err).Warnf("Failed to stat %s", path)
		return
	}

	if info.IsDir() {
		os.Chmod(path, 0755) // drwxr-xr-x for directories
	} else {
		os.Chmod(path, 0644) // -rw-r--r-- for files
	}
}

func handleUpload(c *gin.Context) {
	user := c.MustGet("user").(*auth.User)
	uploadDir := fmt.Sprintf("/tmp/uploads/%s", user.ID)

	if err := os.MkdirAll(uploadDir, 0755); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create upload directory"})
		return
	}
	setContainerOwnership(uploadDir)

	form, err := c.MultipartForm()
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Failed to parse multipart form"})
		return
	}

	files := form.File["files"]
	if len(files) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "No files provided"})
		return
	}

	for _, fileHeader := range files {
		if err := saveUploadedFile(fileHeader, uploadDir); err != nil {
			logrus.WithError(err).Errorf("Failed to save file %s", fileHeader.Filename)
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to save file"})
			return
		}
	}

	c.JSON(http.StatusOK, gin.H{"message": "File uploaded successfully"})
}

func handleUploadFolder(c *gin.Context) {
	user := c.MustGet("user").(*auth.User)
	uploadDir := fmt.Sprintf("/tmp/uploads/%s", user.ID)

	if err := os.MkdirAll(uploadDir, 0755); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create upload directory"})
		return
	}
	setContainerOwnership(uploadDir)

	form, err := c.MultipartForm()
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Failed to parse multipart form"})
		return
	}

	files := form.File["files"]
	if len(files) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "No files provided"})
		return
	}

	for _, fileHeader := range files {
		safePath := filepath.Join(uploadDir, fileHeader.Filename)
		safePath = filepath.Clean(safePath)
		
		// Prevent directory traversal
		if !strings.HasPrefix(safePath, uploadDir) {
			continue
		}

		dirPath := filepath.Dir(safePath)
		if err := os.MkdirAll(dirPath, 0755); err != nil {
			logrus.WithError(err).Errorf("Failed to create directory %s", dirPath)
			continue
		}
		setContainerOwnership(dirPath)

		if err := saveUploadedFile(fileHeader, filepath.Dir(safePath)); err != nil {
			logrus.WithError(err).Errorf("Failed to save file %s", fileHeader.Filename)
			continue
		}
	}

	// Check if move-into parameter is provided
	moveInto := c.PostForm("move-into")
	if moveInto != "" {
		containerName := fmt.Sprintf("user-container-%s", user.ID)
		
		// Send Enter key first
		if err := docker.SendTmuxKeys(containerName, "3compute", "Enter"); err != nil {
			logrus.WithError(err).Warn("Failed to send Enter key to tmux")
		}
		
		// Then send cd command
		cdCommand := fmt.Sprintf("cd /app/%s", moveInto)
		if err := docker.SendTmuxKeys(containerName, "3compute", cdCommand, "Enter"); err != nil {
			logrus.WithError(err).Warn("Failed to send cd command to tmux")
		}
	}

	c.JSON(http.StatusOK, gin.H{"message": "Folder uploaded successfully"})
}

func handleListFiles(c *gin.Context) {
	user := c.MustGet("user").(*auth.User)
	uploadDir := fmt.Sprintf("/tmp/uploads/%s", user.ID)

	if _, err := os.Stat(uploadDir); os.IsNotExist(err) {
		c.JSON(http.StatusOK, gin.H{"files": []string{}})
		return
	}

	var fileTree []string

	err := filepath.Walk(uploadDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		// Skip the root directory itself
		if path == uploadDir {
			return nil
		}

		relativePath, err := filepath.Rel(uploadDir, path)
		if err != nil {
			return err
		}

		if info.IsDir() {
			// Include directories with trailing slash
			fileTree = append(fileTree, relativePath+"/")
		} else {
			// Include files
			fileTree = append(fileTree, relativePath)
		}

		return nil
	})

	if err != nil {
		logrus.WithError(err).Error("Failed to walk directory")
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to list files"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"files": fileTree})
}

func handleMoveFileOrFolder(c *gin.Context) {
	user := c.MustGet("user").(*auth.User)

	var data struct {
		Source      string `json:"source"`
		Destination string `json:"destination"`
		Overwrite   bool   `json:"overwrite"`
	}

	if err := c.ShouldBindJSON(&data); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request data"})
		return
	}

	source := strings.TrimPrefix(data.Source, "/")
	destination := strings.TrimPrefix(data.Destination, "/")

	if source == "" || destination == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid path"})
		return
	}

	uploadDir := fmt.Sprintf("/tmp/uploads/%s", user.ID)
	srcPath := filepath.Join(uploadDir, source)
	dstPath := filepath.Join(uploadDir, destination)

	// Security: ensure resulting paths stay within the user's directory
	srcPath = filepath.Clean(srcPath)
	dstPath = filepath.Clean(dstPath)
	if !strings.HasPrefix(srcPath, uploadDir) || !strings.HasPrefix(dstPath, uploadDir) {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid path"})
		return
	}

	if _, err := os.Stat(srcPath); os.IsNotExist(err) {
		c.JSON(http.StatusNotFound, gin.H{"error": "Source not found"})
		return
	}

	// Disallow moving a folder into itself or its descendants
	srcRel, _ := filepath.Rel(uploadDir, srcPath)
	dstRel, _ := filepath.Rel(uploadDir, dstPath)
	if dstRel == srcRel || strings.HasPrefix(dstRel, srcRel+string(filepath.Separator)) {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Cannot move a folder into itself or its subdirectory"})
		return
	}

	// Ensure destination parent exists
	dstParent := filepath.Dir(dstPath)
	if dstParent != uploadDir {
		if err := os.MkdirAll(dstParent, 0755); err != nil {
			c.JSON(http.StatusConflict, gin.H{"error": "A file exists in the destination path"})
			return
		}
		setContainerOwnership(dstParent)
	}

	// Prevent overwriting existing files/folders unless overwrite flag is set
	if _, err := os.Stat(dstPath); err == nil {
		if !data.Overwrite {
			c.JSON(http.StatusConflict, gin.H{"error": "Destination already exists"})
			return
		}
		// If overwriting, remove existing destination first
		if err := os.RemoveAll(dstPath); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": fmt.Sprintf("Failed to replace destination: %v", err)})
			return
		}
	}

	if err := os.Rename(srcPath, dstPath); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": fmt.Sprintf("Failed to move: %v", err)})
		return
	}

	// Set ownership on the moved item
	setContainerOwnership(dstPath)

	c.JSON(http.StatusOK, gin.H{"message": "Moved successfully"})
}

func handleFile(c *gin.Context) {
	user := c.MustGet("user").(*auth.User)
	filename := strings.TrimPrefix(c.Param("filepath"), "/")
	
	if filename == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "No filename provided"})
		return
	}

	uploadDir := fmt.Sprintf("/tmp/uploads/%s", user.ID)
	filePath := filepath.Join(uploadDir, filename)
	filePath = filepath.Clean(filePath)

	// Security check
	if !strings.HasPrefix(filePath, uploadDir) {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid path"})
		return
	}

	switch c.Request.Method {
	case "POST":
		handleCreateFileOrDir(c, filePath, filename)
	case "GET":
		handleGetFile(c, filePath, filename)
	case "PUT":
		handleUpdateFile(c, filePath)
	case "DELETE":
		handleDeleteFile(c, filePath)
	}
}

func handleCreateFileOrDir(c *gin.Context, filePath, filename string) {
	if strings.HasSuffix(filename, "/") {
		// It's a directory
		normalizedPath := strings.TrimSuffix(filePath, "/")
		
		// If a file exists with the same name, return conflict
		if info, err := os.Stat(normalizedPath); err == nil && !info.IsDir() {
			c.JSON(http.StatusConflict, gin.H{"error": "A file with the same name already exists"})
			return
		}
		
		if err := os.MkdirAll(filePath, 0755); err != nil {
			c.JSON(http.StatusConflict, gin.H{"error": "A file exists in the path; cannot create directory"})
			return
		}
		setContainerOwnership(filePath)
		c.JSON(http.StatusOK, gin.H{"message": "Directory created successfully"})
	} else {
		// It's a file
		dirPath := filepath.Dir(filePath)
		uploadDir := fmt.Sprintf("/tmp/uploads/%s", c.MustGet("user").(*auth.User).ID)
		
		if dirPath != uploadDir {
			if err := os.MkdirAll(dirPath, 0755); err != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create directory"})
				return
			}
			setContainerOwnership(dirPath)
		}

		// If a directory with the same name exists, return conflict
		if info, err := os.Stat(filePath); err == nil && info.IsDir() {
			c.JSON(http.StatusConflict, gin.H{"error": "A folder with the same name already exists"})
			return
		}

		if _, err := os.Stat(filePath); err == nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "File already exists"})
			return
		}

		file, err := os.Create(filePath)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create file"})
			return
		}
		file.Close()
		setContainerOwnership(filePath)
		c.JSON(http.StatusOK, gin.H{"message": "File created successfully"})
	}
}

func handleGetFile(c *gin.Context, filePath, filename string) {
	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		c.JSON(http.StatusNotFound, gin.H{"error": "File not found"})
		return
	}

	// Check if it's an image file
	imageExtensions := []string{"jpg", "jpeg", "png", "gif", "bmp", "webp", "svg", "ico"}
	ext := strings.ToLower(filepath.Ext(filename))
	if len(ext) > 0 {
		ext = ext[1:] // Remove the dot
	}

	isImage := false
	for _, imgExt := range imageExtensions {
		if ext == imgExt {
			isImage = true
			break
		}
	}

	if isImage {
		c.File(filePath)
		return
	}

	// Try to read as text file
	content, err := os.ReadFile(filePath)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to read file"})
		return
	}

	// Check if content is valid UTF-8
	if !isValidUTF8(content) {
		// If not valid UTF-8, serve as binary file
		c.File(filePath)
		return
	}

	c.String(http.StatusOK, string(content))
}

func handleUpdateFile(c *gin.Context, filePath string) {
	content, err := io.ReadAll(c.Request.Body)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Failed to read request body"})
		return
	}

	if err := os.WriteFile(filePath, content, 0644); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to write file"})
		return
	}

	setContainerOwnership(filePath)
	c.JSON(http.StatusOK, gin.H{"message": "File updated successfully"})
}

func handleDeleteFile(c *gin.Context, filePath string) {
	if err := os.RemoveAll(filePath); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete file"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "File deleted successfully"})
}

// saveUploadedFile saves an uploaded file to the specified directory
func saveUploadedFile(fileHeader *multipart.FileHeader, dir string) error {
	src, err := fileHeader.Open()
	if err != nil {
		return err
	}
	defer src.Close()

	filePath := filepath.Join(dir, fileHeader.Filename)
	dst, err := os.Create(filePath)
	if err != nil {
		return err
	}
	defer dst.Close()

	if _, err := io.Copy(dst, src); err != nil {
		return err
	}

	setContainerOwnership(filePath)
	return nil
}

// isValidUTF8 checks if the content is valid UTF-8
func isValidUTF8(content []byte) bool {
	// Simple check - if we can convert to string without replacement characters
	str := string(content)
	return !strings.Contains(str, "\uFFFD")
}
